File: services/aep-api/internal/spec/genaiturns/aim_test.go
// Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
//
// WSO2 LLC. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package genaiturns

import (
	"testing"

	"github.com/wso2/aep/aep-api/internal/gen"
)

func anchorFixture() gen.TurnAnchor {
	return gen.TurnAnchor{
		File: "specs/requirements/PRD.md",
		Nodes: []gen.TurnAnchorNode{
			{Name: "Rounds close automatically", Kind: "paragraph", Context: "Product Decisions"},
		},
	}
}

// An ordinary chat turn must reach the agents service byte-identical to one
// sent before aiming existed — nil, not an empty block.
func TestAimAbsentWhenNeitherFieldSent(t *testing.T) {
	aim, err := aimFromJSON(gen.TurnAnchor{}, "")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if aim != nil {
		t.Fatalf("expected no aim, got %+v", aim)
	}
}

// Both or neither. An intent with nothing to point at says nothing, and an
// anchor with no intent leaves the agents service guessing which preamble to
// render — rejecting here keeps that guess from ever existing.
func TestAimHalfSetIsRejected(t *testing.T) {
	if _, err := aimFromJSON(anchorFixture(), ""); err == nil {
		t.Fatal("anchor without intent should be rejected")
	}
	if _, err := aimFromJSON(gen.TurnAnchor{}, "change"); err == nil {
		t.Fatal("intent without anchor should be rejected")
	}
}

func TestAimCarriesEveryNode(t *testing.T) {
	anchor := anchorFixture()
	anchor.Nodes = append(anchor.Nodes, gen.TurnAnchorNode{Name: "One message per round", Kind: "list item"})
	aim, err := aimFromJSON(anchor, "discuss")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if aim.Intent != "discuss" || aim.Anchor.File != anchor.File {
		t.Fatalf("aim did not survive conversion: %+v", aim)
	}
	if len(aim.Anchor.Nodes) != 2 {
		t.Fatalf("expected 2 nodes, got %d", len(aim.Anchor.Nodes))
	}
	// A name that stands alone carries no context, and must not grow a blank one.
	if aim.Anchor.Nodes[1].Context != "" {
		t.Fatalf("expected no context, got %q", aim.Anchor.Nodes[1].Context)
	}
}

func TestAimRejectsUnknownIntent(t *testing.T) {
	if _, err := aimFromJSON(anchorFixture(), "rewrite"); err == nil {
		t.Fatal("an intent outside change|discuss should be rejected")
	}
}

func TestAimRejectsIncompleteNode(t *testing.T) {
	anchor := anchorFixture()
	anchor.Nodes = append(anchor.Nodes, gen.TurnAnchorNode{Name: "no kind"})
	if _, err := aimFromJSON(anchor, "change"); err == nil {
		t.Fatal("a node without a kind should be rejected")
	}
}

func TestAimRejectsEmptyNodeList(t *testing.T) {
	if _, err := aimFromJSON(gen.TurnAnchor{File: "a.md"}, "change"); err == nil {
		t.Fatal("an anchor with no nodes should be rejected")
	}
}

// Past this many nodes the user is pointing at the document, not a passage, and
// the preamble stops naming anything the agent can act on.
func TestAimRejectsAnUnboundedSelection(t *testing.T) {
	anchor := gen.TurnAnchor{File: "a.md"}
	for i := 0; i <= maxAnchorNodes; i++ {
		anchor.Nodes = append(anchor.Nodes, gen.TurnAnchorNode{Name: "n", Kind: "paragraph"})
	}
	if _, err := aimFromJSON(anchor, "change"); err == nil {
		t.Fatal("a selection past the node cap should be rejected")
	}
}

// The anchor locates and never carries (console ADR-0024): a hand-built
// request must not smuggle a prompt payload through the locator fields the
// console keeps short by construction.
func TestAimRejectsOversizedFields(t *testing.T) {
	long := func(n int) string {
		b := make([]byte, n)
		for i := range b {
			b[i] = 'x'
		}
		return string(b)
	}
	oversized := []gen.TurnAnchor{
		{File: long(513), Nodes: []gen.TurnAnchorNode{{Name: "n", Kind: "paragraph"}}},
		{File: "a.md", Nodes: []gen.TurnAnchorNode{{Name: long(201), Kind: "paragraph"}}},
		{File: "a.md", Nodes: []gen.TurnAnchorNode{{Name: "n", Kind: long(65)}}},
		{File: "a.md", Nodes: []gen.TurnAnchorNode{{Name: "n", Kind: "paragraph", Context: long(513)}}},
	}
	for i, anchor := range oversized {
		if _, err := aimFromJSON(anchor, "change"); err == nil {
			t.Fatalf("case %d: an oversized anchor field should be rejected", i)
		}
	}
	// At the ceiling is fine — the cap is a ceiling, not a headroom rule.
	atLimit := gen.TurnAnchor{File: "a.md", Nodes: []gen.TurnAnchorNode{{Name: long(200), Kind: "paragraph"}}}
	if _, err := aimFromJSON(atLimit, "change"); err != nil {
		t.Fatalf("an anchor at the ceiling should pass: %v", err)
	}
}

// The multipart form declares this part application/json: a nested object has
// no scalar form, unlike `collab`, which rides as the string "true".
func TestAnchorFieldDecodesTheJSONPart(t *testing.T) {
	anchor, err := parseAnchorField(`{"file":"a.md","nodes":[{"name":"n","kind":"paragraph"}]}`)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if anchor.File != "a.md" || len(anchor.Nodes) != 1 {
		t.Fatalf("anchor did not decode: %+v", anchor)
	}
}

func TestAnchorFieldRejectsMalformedJSON(t *testing.T) {
	if _, err := parseAnchorField("{not json"); err == nil {
		t.Fatal("malformed anchor JSON should be rejected")
	}
}

// An absent part is absent, not an error: a multipart send with attachments and
// no selection is an ordinary turn.
func TestAnchorFieldAbsentIsNotAnError(t *testing.T) {
	anchor, err := parseAnchorField("")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if anchor.File != "" || len(anchor.Nodes) != 0 {
		t.Fatalf("expected a zero anchor, got %+v", anchor)
	}
}

File: services/agents/test/aim.test.ts
/**
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { composeInstruction, aimNote } from "../src/prompts/turn.js";
import { isTurnAim, type TurnAim } from "@aep/agent-stream";
import { projectDisplayHistory } from "../src/conversation/display-history.js";
import type { Conversation } from "../src/store/conversation-store.js";

// An anchored turn (console #666): the user pointed at part of a spec document
// before typing. What is proven here is that the anchor becomes WORDS in this
// service and nowhere else — the console sends the user's own text plus a fact,
// and the preamble is composed here, because a preamble written client-side
// would have to ride `instruction` and would then appear in the transcript as
// something the user said and did not.

const AIM: TurnAim = {
  anchor: {
    file: "specs/requirements/PRD.md",
    nodes: [
      {
        name: "Rounds close automatically thirty minutes before the delivery slot",
        kind: "paragraph",
        context: "Product Decisions",
      },
      { name: "The bot posts one message per round", kind: "list item" },
    ],
  },
  intent: "change",
};

test("an unaimed turn is byte-identical to one from before aiming existed", () => {
  const turn = { kind: "chat", text: "tidy the requirements" } as const;
  assert.equal(composeInstruction(turn), composeInstruction(turn, {}));
  assert.equal(aimNote(undefined), "");
});

test("the aim leads the instruction, so the model knows what it is pointed at first", () => {
  const composed = composeInstruction({ kind: "chat", text: "make these shorter" }, { aim: AIM });
  assert.ok(composed.indexOf("selected part of a document") < composed.indexOf("make these shorter"));
});

test("every selected node is named, with its kind and where it sits", () => {
  const note = aimNote(AIM);
  assert.match(note, /specs\/requirements\/PRD\.md/);
  assert.match(note, /- the paragraph "Rounds close automatically thirty minutes before the delivery slot" \(under Product Decisions\)/);
  // A name that stands alone carries no context, and must not grow an empty one.
  assert.match(note, /- the list item "The bot posts one message per round"\n/);
  assert.doesNotMatch(note, /\(under \)/);
});

// The decision this test exists for (console ADR-0024). The names are LOCATORS,
// not content: the agent is a live peer in the same room, so the document may
// have moved since the user selected. An agent that guesses which paragraph was
// meant is the one failure the whole shape exists to avoid.
test("the note says to resolve against the current document, and to ask on a miss", () => {
  const note = aimNote(AIM);
  assert.match(note, /as it stands now/);
  assert.match(note, /never guess/);
});

// The fence (#654): the selection is the SUBJECT, not a cage. A spec has
// legitimate ripples, so a change may touch what it makes wrong — and nothing
// else. When the right fix lies elsewhere, the agent makes it there and says
// so, because the user asked for a change, not a conversation — Discuss is
// the asking path.
test("a change is fenced to its subject, with consistency allowed and tidying forbidden", () => {
  const note = aimNote(AIM);
  assert.match(note, /the subject, not a cage/);
  assert.match(note, /consistency, never improvement/);
  assert.match(note, /make it there and say so plainly/);
});

test("a discuss carries no fence — it edits nothing to fence", () => {
  assert.doesNotMatch(aimNote({ ...AIM, intent: "discuss" }), /not a cage/);
});

test("discuss and change differ, and only in the wording", () => {
  const change = aimNote(AIM);
  const discuss = aimNote({ ...AIM, intent: "discuss" });
  assert.notEqual(change, discuss);
  assert.match(discuss, /do not edit the document in this turn/i);
  assert.doesNotMatch(change, /do not edit the document in this turn/i);
  // The selection itself reads the same either way — the intent changes what to
  // do with it, never what was pointed at.
  for (const node of AIM.anchor.nodes) {
    assert.ok(change.includes(node.name));
    assert.ok(discuss.includes(node.name));
  }
});

// The guard is what stands between an untrusted body and a preamble that points
// the agent at the wrong scope quietly. A half-read aim is worse than none.
// The caps (CodeRabbit on #670): the anchor locates and never carries, so a
// hand-built request must not smuggle a prompt payload through the locator
// fields the console keeps short by construction.
test("an aim past the size ceilings is rejected", () => {
  const long = (n: number) => "x".repeat(n);
  assert.ok(!isTurnAim({ ...AIM, anchor: { ...AIM.anchor, file: long(513) } }));
  assert.ok(
    !isTurnAim({
      ...AIM,
      anchor: { file: "a.md", nodes: [{ name: long(201), kind: "paragraph" }] },
    }),
  );
  assert.ok(
    !isTurnAim({
      ...AIM,
      anchor: { file: "a.md", nodes: [{ name: "n", kind: long(65) }] },
    }),
  );
  assert.ok(
    !isTurnAim({
      ...AIM,
      anchor: { file: "a.md", nodes: [{ name: "n", kind: "paragraph", context: long(513) }] },
    }),
  );
  assert.ok(
    !isTurnAim({
      ...AIM,
      anchor: {
        file: "a.md",
        nodes: Array.from({ length: 51 }, () => ({ name: "n", kind: "paragraph" })),
      },
    }),
  );
  // At the ceiling is fine — the cap is a ceiling, not a headroom rule.
  assert.ok(
    isTurnAim({
      ...AIM,
      anchor: { file: "a.md", nodes: [{ name: long(200), kind: "paragraph" }] },
    }),
  );
});

test("a malformed aim is rejected whole", () => {
  assert.ok(isTurnAim(AIM));
  assert.ok(!isTurnAim({ ...AIM, intent: "rewrite" }));
  assert.ok(!isTurnAim({ anchor: AIM.anchor }));
  assert.ok(!isTurnAim({ ...AIM, anchor: { file: "", nodes: AIM.anchor.nodes } }));
  assert.ok(!isTurnAim({ ...AIM, anchor: { file: "a.md", nodes: [] } }));
  assert.ok(
    !isTurnAim({ ...AIM, anchor: { file: "a.md", nodes: [{ name: "n" }] } }),
    "a node with no kind",
  );
});

// Without this a reload leaves "make these shorter" with nothing saying what
// "these" was — a record of the words but not the target.
test("the journalled anchor is served back on the display read", () => {
  const conv = {
    messages: [
      { role: "user", content: "composed prompt the browser must never see" },
      { role: "assistant", content: "done" },
    ],
    turns: [
      {
        turnId: "t-1",
        text: "make these shorter",
        anchor: AIM.anchor,
        messageIndex: 0,
        createdAt: new Date(),
      },
    ],
  } as unknown as Conversation;

  const [user] = projectDisplayHistory(conv);
  assert.equal(user?.role, "user");
  assert.equal((user as { content: string }).content, "make these shorter");
  assert.deepEqual((user as { anchor?: unknown }).anchor, AIM.anchor);
});

test("an un-anchored journal entry grows no anchor field", () => {
  const conv = {
    messages: [{ role: "user", content: "composed" }],
    turns: [{ turnId: "t-1", text: "hello", messageIndex: 0, createdAt: new Date() }],
  } as unknown as Conversation;
  assert.ok(!("anchor" in (projectDisplayHistory(conv)[0] as object)));
});
